// trm-core/dist/actions/checkSapEntries/buildOutput.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildOutput = void 0;
exports.buildOutput = {
    name: 'build-output',
    filter: (context) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            const items = context.runtime.okEntries.length +
                context.runtime.okEntries.length;
            return items > 0;
        }
        catch (e) {
            return false;
        }
    }),
    run: (context) => __awaiter(void 0, void 0, void 0, function* () {
        context.output.sapEntriesStatus = {};
        context.runtime.okEntries.forEach(o => {
            if (!context.output.sapEntriesStatus[o.table]) {
                context.output.sapEntriesStatus[o.table] = [];
            }
            context.output.sapEntriesStatus[o.table].push({
                status: true,
                entry: o.tableEntry
            });
        });
        context.runtime.koEntries.forEach(o => {
            if (!context.output.sapEntriesStatus[o.table]) {
                context.output.sapEntriesStatus[o.table] = [];
            }
            context.output.sapEntriesStatus[o.table].push({
                status: false,
                entry: o.tableEntry
            });
        });
    })
};
