// trm-core/dist/logger/ConsoleLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleLogger = void 0;
const trm_registry_types_1 = require("trm-registry-types");
class ConsoleLogger {
    constructor(debug) { }
    loading(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.log(text);
    }
    success(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.log(text);
    }
    error(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.error(text);
    }
    warning(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.warn(text);
    }
    info(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.info(text);
    }
    log(text, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.log(text);
    }
    table(header, data, debug) {
        if (debug && !this.debug) {
            return;
        }
        const table = {
            header,
            data
        };
        console.log(JSON.stringify(table));
    }
    registryResponse(response, debug) {
        if (debug && !this.debug) {
            return;
        }
        if (response.type === trm_registry_types_1.MessageType.ERROR) {
            this.error(response.text, debug);
        }
        if (response.type === trm_registry_types_1.MessageType.INFO) {
            this.info(response.text, debug);
        }
        if (response.type === trm_registry_types_1.MessageType.WARNING) {
            this.warning(response.text, debug);
        }
    }
    tree(data, debug) {
        if (debug && !this.debug) {
            return;
        }
        console.log(JSON.stringify(data));
    }
}
exports.ConsoleLogger = ConsoleLogger;
