// trm-core/dist/actions/publish/editTrmDependencies.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.editTrmDependencies = void 0;
const logger_1 = require("../../logger");
const Inquirer_1 = require("../../inquirer/Inquirer");
exports.editTrmDependencies = {
    name: 'edit-trm-dependencies',
    filter: (context) => __awaiter(void 0, void 0, void 0, function* () {
        if (context.parsedInput.skipEditDependencies || context.parsedInput.silent) {
            logger_1.Logger.log(`Skip edit of TRM dependencies (input)`, true);
            return false;
        }
        else {
            return true;
        }
    }),
    run: (context) => __awaiter(void 0, void 0, void 0, function* () {
        var editorValue = '[]';
        if (context.runtime.manifest.dependencies) {
            editorValue = JSON.stringify(context.runtime.manifest.dependencies, null, 2);
        }
        const inq1 = yield Inquirer_1.Inquirer.prompt([{
                message: `Manually edit dependencies?`,
                type: 'confirm',
                name: 'editDependencies',
                default: false
            }, {
                message: 'Editor dependencies',
                type: 'editor',
                name: 'dependencies',
                postfix: '.json',
                when: (hash) => {
                    return hash.editDependencies;
                },
                default: editorValue,
                validate: (input) => {
                    try {
                        const parsedInput = JSON.parse(input);
                        if (Array.isArray(parsedInput)) {
                            return true;
                        }
                        else {
                            return 'Invalid array';
                        }
                    }
                    catch (e) {
                        return 'Invalid JSON';
                    }
                }
            }]);
        if (inq1.dependencies) {
            logger_1.Logger.log(`TRM dependencies have been manually edited: before -> ${JSON.stringify(context.runtime.manifest.dependencies)}, after -> ${inq1.dependencies}`, true);
            context.runtime.manifest.dependencies = JSON.parse(inq1.dependencies);
        }
        else {
            logger_1.Logger.log(`TRM dependencies are not been manually edited`, true);
            context.runtime.manifest.dependencies = context.runtime.manifest.dependencies || [];
        }
    })
};
