// trm-core/dist/actions/publish/generateDevcTr.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDevcTr = void 0;
const transport_1 = require("../../transport");
const logger_1 = require("../../logger");
exports.generateDevcTr = {
    name: 'generate-devc-tr',
    run: (context) => __awaiter(void 0, void 0, void 0, function* () {
        logger_1.Logger.loading(`Generating DEVC transport...`);
        const devcOnly = context.runtime.tadirObjects.filter(o => o.pgmid === 'R3TR' && o.object === 'DEVC');
        context.runtime.devcTransport = yield transport_1.Transport.createToc({
            trmIdentifier: transport_1.TrmTransportIdentifier.DEVC,
            target: context.parsedInput.trTarget,
            text: `@X1@TRM: ${context.runtime.manifest.name} v${context.runtime.manifest.version} (D)`
        });
        context.runtime.tryDevcDeleteRevert = true;
        yield context.runtime.devcTransport.addObjects(devcOnly, false);
    }),
    revert: (context) => __awaiter(void 0, void 0, void 0, function* () {
        if (context.runtime.tryDevcDeleteRevert) {
            logger_1.Logger.loading(`Rollback DEVC transport ${context.runtime.devcTransport.trkorr}...`);
            try {
                const canBeDeleted = yield context.runtime.devcTransport.canBeDeleted();
                if (canBeDeleted) {
                    yield context.runtime.devcTransport.delete();
                    logger_1.Logger.info(`Executed rollback on transport ${context.runtime.devcTransport.trkorr}`);
                }
                else {
                    throw new Error(`Transport ${context.runtime.devcTransport.trkorr} cannot be deleted`);
                }
            }
            catch (e) {
                logger_1.Logger.info(`Unable to rollback transport ${context.runtime.devcTransport.trkorr}`);
                logger_1.Logger.error(e.toString(), true);
            }
        }
    })
};
