// trm-core/dist/actions/publish/generateCustTr.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCustTr = void 0;
const transport_1 = require("../../transport");
const logger_1 = require("../../logger");
exports.generateCustTr = {
    name: 'generate-cust-tr',
    filter: (context) => __awaiter(void 0, void 0, void 0, function* () {
        if (context.runtime.inputCustTransports.length > 0) {
            return true;
        }
        else {
            logger_1.Logger.log(`Skipping CUST transport (no input in step)`, true);
            return false;
        }
    }),
    run: (context) => __awaiter(void 0, void 0, void 0, function* () {
        const customizingTransports = context.runtime.inputCustTransports;
        logger_1.Logger.loading(`Generating CUST transport...`);
        context.runtime.custTransport = yield transport_1.Transport.createToc({
            target: context.parsedInput.trTarget,
            text: `@X1@TRM: ${context.runtime.manifest.name} v${context.runtime.manifest.version} (C)`,
            trmIdentifier: transport_1.TrmTransportIdentifier.CUST
        });
        context.runtime.tryCustDeleteRevert = true;
        for (const transport of customizingTransports) {
            yield context.runtime.custTransport.addObjectsFromTransport(transport.trkorr);
        }
        var e071 = yield context.runtime.custTransport.getE071();
        e071 = e071.filter(o => !(o.pgmid === 'CORR' && o.object === 'MERG'));
        if (e071.length === 0) {
            logger_1.Logger.info(`Customizing transport has no content.`);
            yield context.runtime.custTransport.delete();
            delete context.runtime.custTransport;
            context.runtime.tryCustDeleteRevert = false;
        }
        else {
            const wrongEntries = e071.filter(o => !(o.pgmid === 'R3TR' && o.object === 'TABU'));
            if (wrongEntries.length > 0) {
                throw new Error(`Customizing transport contains invalid objects.`);
            }
        }
    }),
    revert: (context) => __awaiter(void 0, void 0, void 0, function* () {
        if (context.runtime.tryCustDeleteRevert) {
            logger_1.Logger.loading(`Rollback CUST transport ${context.runtime.custTransport.trkorr}...`);
            try {
                const canBeDeleted = yield context.runtime.custTransport.canBeDeleted();
                if (canBeDeleted) {
                    yield context.runtime.custTransport.delete();
                    logger_1.Logger.info(`Executed rollback on transport ${context.runtime.custTransport.trkorr}`);
                }
                else {
                    throw new Error(`Transport ${context.runtime.custTransport.trkorr} cannot be deleted`);
                }
            }
            catch (e) {
                logger_1.Logger.info(`Unable to rollback transport ${context.runtime.custTransport.trkorr}`);
                logger_1.Logger.error(e.toString(), true);
            }
        }
    })
};
