// trm-core/dist/actions/publish/generateTadirTr.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateTadirTr = void 0;
const transport_1 = require("../../transport");
const logger_1 = require("../../logger");
const systemConnector_1 = require("../../systemConnector");
exports.generateTadirTr = {
    name: 'generate-tadir-tr',
    run: (context) => __awaiter(void 0, void 0, void 0, function* () {
        logger_1.Logger.loading(`Generating TADIR transport...`);
        const sManifestXml = context.runtime.trmPackage.manifest.getAbapXml();
        const objectsOnly = context.runtime.tadirObjects.filter(o => !(o.pgmid === 'R3TR' && o.object === 'DEVC'));
        context.runtime.tadirTransport = yield transport_1.Transport.createToc({
            trmIdentifier: transport_1.TrmTransportIdentifier.TADIR,
            target: context.parsedInput.trTarget,
            text: `@X1@TRM: ${context.runtime.manifest.name} v${context.runtime.manifest.version}`
        });
        context.runtime.tryTadirDeleteRevert = true;
        yield context.runtime.tadirTransport.addComment(`name=${context.runtime.manifest.name}`);
        yield context.runtime.tadirTransport.addComment(`version=${context.runtime.manifest.version}`);
        yield context.runtime.tadirTransport.setDocumentation(sManifestXml);
        yield context.runtime.tadirTransport.addObjects(objectsOnly, false);
    }),
    revert: (context) => __awaiter(void 0, void 0, void 0, function* () {
        if (context.runtime.tryTadirDeleteRevert) {
            logger_1.Logger.loading(`Rollback TADIR transport ${context.runtime.tadirTransport.trkorr}...`);
            try {
                const canBeDeleted = yield context.runtime.tadirTransport.canBeDeleted();
                if (canBeDeleted) {
                    yield context.runtime.tadirTransport.delete();
                }
                else {
                    yield systemConnector_1.SystemConnector.addSkipTrkorr(context.runtime.tadirTransport.trkorr);
                }
                logger_1.Logger.info(`Executed rollback on transport ${context.runtime.tadirTransport.trkorr}`);
            }
            catch (e) {
                logger_1.Logger.info(`Unable to rollback transport ${context.runtime.tadirTransport.trkorr}`);
                logger_1.Logger.error(e.toString(), true);
            }
        }
    })
};
